timelock-impl/src/test/java/com/palantir/atlasdb/timelock/transaction/client/DistributingModulusGeneratorTest.java

```diff
@@ -0,0 +1,164 @@
+        setupGeneratorWithModulus(1);
+
+        IntStream.range(0, 50).forEach(unused -> requestResiduesAndAssertBalanced(1));
+    }
+
+    @Test
+    public void throwsIfCreatedWithNonPositiveModuli() {
+        assertThatThrownBy(() -> setupGeneratorWithModulus(0))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("Modulus must be positive");
+        assertThatThrownBy(() -> setupGeneratorWithModulus(-8))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("Modulus must be positive");
+    }
+
+    @Test
+    public void throwsIfUnmarkingUnmarkedResidue() {
+        setupGeneratorWithModulus(3);
+
+        assertThatThrownBy(() -> unmarkResidue(0)).satisfies(this::failureArisesFromUnmarkingResidues);
+    }
+
+    @Test
+    public void canUnmarkMarkedResidues() {
+        setupGeneratorWithModulus(3);
+
+        requestResidues(3);
+        unmarkResidue(0);
+        unmarkResidue(1);
+        unmarkResidue(2);
+        // Not throwing is a success!
+    }
+
+    @Test
+    public void cannotDoubleUnmarkMarkedResidues() {
+        setupGeneratorWithModulus(3);
+
+        requestResidues(3);
+
+        unmarkResidue(0);
+        assertThatThrownBy(() -> unmarkResidue(0)).satisfies(this::failureArisesFromUnmarkingResidues);
+    }
+
+    @Test
+    public void reassignsUnmarkedResiduesIfTheyAreNowLeastReferenced() {
+        setupGeneratorWithModulus(3);
+
+        requestResiduesAndAssertBalanced(3);
+        unmarkResidue(1);
+        assertThat(requestResidues(1)).containsExactly(1);
+    }
+
+    @Test
+    public void repeatedlyReassignsUnmarkedResidues() {
+        setupGeneratorWithModulus(3);
+
+        requestResiduesAndAssertBalanced(3);
+        requestResiduesAndAssertBalanced(3);
+        requestResiduesAndAssertBalanced(3);
+
+        unmarkResidue(2);
+        unmarkResidue(2);
+        unmarkResidue(2);
+
+        List<Integer> additionalResponses = requestResidues(3);
+        assertThat(additionalResponses).containsExactly(2, 2, 2);
+
+        requestResiduesAndAssertBalanced(3);
+    }
+
+    private void setupGeneratorWithModulus(int modulus) {
+        generator = new DistributingModulusGenerator(modulus);
+    }
+
+    private List<Integer> requestResidues(int times) {
+        return IntStream.range(0, times)
+                .map(unused -> generator.getAndMarkResidue())
+                .boxed()
+                .collect(Collectors.toList());
+    }
+
+    private void requestResiduesAndAssertBalanced(int times) {
+        assertListBalanced(requestResidues(times), times);
+    }
+
+    /**
+     * Asserts that, starting from the beginning of the list, each full modulus number of elements has the integers
+     * from 0 to (modulus - 1) once each. No guarantees are made on the remainder of the list.
+     */
+    private void assertListBalanced(List<Integer> responses, int modulus) {
```

Comment:
See above comment about being verbose, i'm not a huge fan of "computations" in tests, as a reviewer I have to effectively run through in my head what potential outputs this should have. This is compounded by javadoc explaining the method.

Also, `static`.

Reply:
I've removed this code since the imbalanced case wasn't actually ever asserted (i.e. we didn't need to say that [0, 1, 2, {one of them}] needs to pass anywhere)